sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java

```diff
@@ -207,6 +206,29 @@ public ShardCollectSource(Settings settings,
         indicesLifecycle.addListener(new LifecycleListener());
     }
 
+    /**
+     * Executor that delegates the execution of tasks to the provided {@link Executor} until
+     * it starts rejecting tasks with {@link EsRejectedExecutionException} in which case it dispatches the
+     * rejected task to a DirectExecutor
+     */
+    public static class DispatchingExecutor implements Executor {
+
+        private Executor delegateExecutor;
+
+        public DispatchingExecutor(Executor delegateExecutor) {
+            this.delegateExecutor = delegateExecutor;
+        }
+
+        @Override
+        public void execute(Runnable command) {
+            try {
+                delegateExecutor.execute(command);
+            } catch (EsRejectedExecutionException e) {
+                MoreExecutors.directExecutor().execute(command);
```

Comment:
Could also call `command.run()` directly instead of going through the `directExecutor()` indirection.